app/models/assignment_generator.rb

```diff
@@ -0,0 +1,308 @@
+# minimizing the number of assignments with no recipient or no giver). Sorts
+# the potential matches (using the default sort, which relies on the fields
+# num_prompts_matched and max_tags_matched) and uses that order to process
+# outgoing edges in the depth-first search, so that we prioritize "good"
+# matches.
+#
+# Follows the pseudocode from Wikipedia:
+# https://en.wikipedia.org/wiki/Hopcroft-Karp_algorithm
+#
+# This class also has some extra steps to try to reduce the number of "cycles"
+# (where person A is assigned to write for person B, and person B is assigned
+# to write for person A). It's not guaranteed to be the minimum possible, but
+# it might make the assignments nicer.
+class AssignmentGenerator
+  # Save the collection we're working with.
+  def initialize(collection)
+    @collection = collection
+    @settings = @collection.challenge.potential_match_settings
+  end
+
+  # Generate all assignments for the desired collection.
+  def generate
+    load_potential_matches
+
+    set_up_assignments
+
+    while improve_assignments
+      # Do nothing, just continue improving.
+    end
+
+    reduce_cycles
+
+    raise "Invalid assignments!" unless validate_assignments
+
+    save_assignments
+    save_placeholder_assignments
+  end
+
+  private
+
+  # Load information about potential matches from the collection.
+  def load_potential_matches
+    potential_matches = if @settings.nil? || @settings.no_match_required?
+                          # Randomize if we have unconstrained matching.
+                          @collection.potential_matches.all.shuffle
+                        else
+                          # Otherwise, order by potential match "goodness",
+                          # highest to lowest, so that we prioritize better
+                          # matches.
+                          @collection.potential_matches.all.sort.reverse
+                        end
+
+    # Load the list of offers associated with each signup. Because we sorted the
+    # potential matches list, each individual list of offers will be ordered by
+    # match "goodness."
+    @potential_givers = {}
+    potential_matches.each do |pm|
+      @potential_givers[pm.request_signup_id] ||= []
+      @potential_givers[pm.request_signup_id] << pm.offer_signup_id
+    end
+
+    # Get a list of all signups that are matchable as a recipient, ordered by
+    # the number of potential matches (in increasing order). This way, we
+    # prioritize recipients who don't have very many people who could write for
+    # them. Also introduce a random element to break ties.
+    @recipients_by_priority = @potential_givers.keys.sort_by do |key|
+      [@potential_givers[key].size, rand]
+    end
+  end
+
+  # Set up the hashtables we use to store our current assignments.
+  def set_up_assignments
+    # This table will contain a mapping from each request to the offer
+    # that they've been (temporarily) assigned to.
+    @assignment_as_recipient = {}
```

Comment:
I feel like "as" makes it hard to parse if recipient is the key or the value. `assignment_by_recipient` maybe? I could be convinced otherwise.

Ditto `assignment_by_giver`.

Reply:
Hmm. I'm not sure if I like "by" here. I feel like if `giver` was assigned to `recip`, I'd expect `assignment_by_recipient[recip]` and `assignment_by_giver[giver]` to return the same object (with information about both the recipient and the giver).

What about `assignment_for_recipient` and `assignment_for_giver`?

Reply:
`_for_` sounds good to me.